_LowLevelCode/CPP/cpp_communicator/MPI_wrapper.h

```diff
@@ -1,25 +1,114 @@
 #pragma once
 #include <vector>
+#include <list>
 #include <cmath>
 #include <mpi.h>
+#include "input_parser.h"
+
+/** Helper class to keep information on send message unit MPI framework reports delivered.
+*
+* in test mode also used to simulate send/receive operations.
+*/
+class SendMessHolder {
+public:
+    // Holder for ISend request property
+    MPI_Request theRequest;
+    // The holder for the tag of the message to send
+    int mess_tag;
+    // The holder for the address of the message to send
+    int destination;
+    // vector of the message contents, used as the buffer of the message contents until the message is received
+    std::vector<uint8_t> mess_body;
+      //SendMessHolder(SendMessHolder&& other) noexcept;
+    SendMessHolder() :
+        mess_tag(-1), destination(-1) {
+        this->theRequest = (MPI_Request)(-1);
+    }
+    SendMessHolder(uint8_t* pBuffer, size_t n_bytes, int dest_address, int data_tag);
+
+    void init(uint8_t* pBuffer, size_t n_bytes, int dest_address, int data_tag);
+};
 
 /* The class which describes a block of information necessary to process block of pixels */
 class MPI_wrapper {
 public:
 
-	MPI_wrapper():
-        labIndex(-1), numProcs(0){}
-	int init();
-	void close();
+    MPI_wrapper() :
+        labIndex(-1), numProcs(0), isTested(false),
+        assynch_queue_max_len_(10) {}
+    int init(bool isTested = false, int assynch_queue_max_len = 10, int data_mess_tag=5);
+    void close();
     void barrier();
+    void clearAll();
+    void labSend(int data_address, int data_tag, bool is_synchroneous, uint8_t* data_buffer, size_t nbytes_to_transfer);
+    void labProbe(const std::vector<int32_t> &data_address, const std::vector<int32_t> &data_tag, 
+        std::vector<int32_t> & addres_present, std::vector<int32_t> & tag_present);
+    void labReceive(int source_address, int source_data_tag, bool isSynchronous, mxArray* plhs[], int nlhs);
+    ~MPI_wrapper() {
+        this->close();
+    }
+    // index of the current MPI lab (worker)
+    int labIndex;
+    // total  number of MPI labs (workers)
+    int numProcs;
+    // test mode used to run various test operations over MPI_wrapper in single process, 
+    // when no real mpi exchange is initiated.
+    bool isTested;
+    // return the number of asynchroneous messages in the queue
+    size_t assync_queue_len() {
+        return this->assyncMessList.size();
+    }
+    // the tag of message, containing data (processed differently, not yet implemented.)
+    static int data_mess_tag;
 
-	~MPI_wrapper() {
-		this->close();
-	}
+    //----------------------------------------------------------------------------------
+    // The methods used in unit tests -- have no meaning in real communications
```

Comment:
if they're only used in unit tests should the be part of the PUBLIC API?

Reply:
Well, I can make it protected and expose it on tester class, but would it be an overkill? 
API is not that extensive and will not be much more complicated (I think two additional public methods are necessary i.e. serialize/deserialize) and everything is clear (hopefully) as it is. 

"Make an program interface a fully fool-proof and only full will use it...."

Reply:
I think, wrapping test code in a test code would be overkill in this situation